sklearn/utils/validation.py

```diff
@@ -397,7 +397,7 @@ def _num_samples(x):
     if hasattr(x, "shape") and x.shape is not None:
         if len(x.shape) == 0:
             raise TypeError(
-                "Singleton array %r cannot be considered a valid collection." % x
+                f"Expected array-like of collection type, got type `{type(x)}` instead."
```

Comment:
I'm not really sure how else to phrase it then "array-like of collection type" (which is a bit of non-sense), because `_num_samples()` is expecting a collection type, `indexable()` a list (which is a collection type), but `check_array()` can deal with almost anything, especially usual arrays.